File: frontend/src/lib/redirects.ts
/**
 * Frontend redirect utilities that mirror backend rules.
 *
 * This module provides redirect safety utilities that enforce:
 * - Single-decode enforcement (decode at most twice)
 * - Prevention of auth page redirects
 * - Same-origin relative paths only
 * - Strip fragments (#...), collapse //, remove nested ?next=...
 * - gs_next cookie support for post-login targets
 */

// Auth paths that should never be redirected to
const AUTH_PATHS = new Set([
    '/login',
    '/v1/auth/login',
    '/v1/auth/logout',
    '/v1/auth/refresh',
    '/v1/auth/csrf',
    '/v1/csrf',
    '/google',
    '/oauth',
    '/sign-in',
    '/sign-up'
]);

// Default fallback path
export const DEFAULT_REDIRECT = '/dashboard';

export function isAuthPath(path: string): boolean {
    if (!path) return false;

    // Check exact matches
    if (AUTH_PATHS.has(path)) return true;

    // Check if path contains auth patterns
    for (const authPath of AUTH_PATHS) {
        if (path.includes(authPath)) return true;
    }

    return false;
}

export function safeDecodeUrl(url: string, maxDecodes: number = 2): string {
    let decoded = url;
    let previous = url;

    for (let i = 0; i < maxDecodes; i++) {
        try {
            previous = decoded;
            decoded = decodeURIComponent(decoded);

            // Stop if no change (no more encoding layers)
            if (decoded === previous) {
                break;
            }
        } catch {
            // If decoding fails at any point, use the last successfully decoded version
            decoded = previous;
            break;
        }
    }

    return decoded;
}

export function sanitizeNextPath(rawPath: string | null | undefined): string {
    if (!rawPath || typeof rawPath !== 'string') {
        return DEFAULT_REDIRECT;
    }

    let path = rawPath.trim();
    if (!path) return DEFAULT_REDIRECT;

    try {
        // Step 1: Safe URL decoding (at most twice)
        path = safeDecodeUrl(path, 2);

        // Step 2: Reject absolute URLs to prevent open redirects
        if (path.startsWith('http://') || path.startsWith('https://')) {
            console.warn('Rejected absolute URL redirect:', path);
            return DEFAULT_REDIRECT;
        }

        // Step 3: Reject protocol-relative URLs (starting with // but not ///)
        if (path.startsWith('//') && !path.startsWith('///')) {
            console.warn('Rejected protocol-relative URL redirect:', path);
            return DEFAULT_REDIRECT;
        }

        // Step 4: Ensure path starts with /
        if (!path.startsWith('/')) {
            console.warn('Rejected non-relative path redirect:', path);
            return DEFAULT_REDIRECT;
        }

        // Step 4.1: Reject obvious local filesystem paths or file downloads
        const lower = path.toLowerCase();
        if (/^\/(var|etc|usr|private|system|library|applications|volumes)\b/.test(lower)) {
            console.warn('Rejected filesystem-like redirect path:', path);
            return DEFAULT_REDIRECT;
        }
        if (/\.(png|jpe?g|gif|webp|svg|pdf|zip|dmg|mov|mp4|mp3|wav|heic|heif)(?:[?#].*)?$/i.test(path)) {
            console.warn('Rejected file-like redirect path:', path);
            return DEFAULT_REDIRECT;
        }
        if (/[\s:]/.test(path)) {
            console.warn('Rejected suspicious redirect path with spaces/colon:', path);
            return DEFAULT_REDIRECT;
        }

        // Step 5: Strip fragments (#...)
        if (path.includes('#')) {
            path = path.split('#')[0];
        }

        // Step 6: Remove any nested ?next=... parameters
        if (path.includes('?')) {
            try {
                const url = new URL(path, 'http://localhost'); // Dummy base for parsing
                url.searchParams.delete('next');
                path = url.pathname + (url.search ? url.search : '');
            } catch {
                // If URL parsing fails, continue with original path
            }
        }

        // Step 7: Prevent redirect loops by blocking auth-related paths
        if (isAuthPath(path)) {
            console.warn('Rejected auth path redirect:', path);
            return DEFAULT_REDIRECT;
        }

        // Step 8: Normalize redundant slashes
        path = path.replace(/\/+/g, '/');

        // Step 9: Basic path validation (no .. traversal)
        if (path.includes('..')) {
            console.warn('Rejected path traversal redirect:', path);
            return DEFAULT_REDIRECT;
        }

        return path;

    } catch (e) {
        console.error('Error sanitizing redirect path', rawPath, e);
        return DEFAULT_REDIRECT;
    }
}

// Helper function to capture next path to backend gs_next cookie
export async function captureNextPathToBackend(nextPath: string): Promise<void> {
    if (!nextPath || nextPath === DEFAULT_REDIRECT) {
        return;
    }

    try {
        // Import apiFetch dynamically to avoid circular dependencies
        const { apiFetch } = await import('@/lib/api');

        // Make a fire-and-forget POST to the backend login endpoint
        // This sets the gs_next cookie on the backend for post-login redirect
        const response = await apiFetch('/v1/auth/login', {
            method: 'POST',
            headers: {
                'Content-Type': 'application/json',
            },
            body: JSON.stringify({
                username: '__capture_next__', // pragma: allowlist secret - Special marker for next path capture
                password: '__capture_next__', // pragma: allowlist secret
                next: nextPath,
            }),
        });

        if (!response.ok) {
            console.warn('Failed to capture next path to backend:', response.status);
        }
    } catch (error) {
        console.warn('Error capturing next path to backend:', error);
    }
}

File: app/security/redirects.py
"""
Redirect sanitization module for secure redirect handling.

This module provides canonical redirect safety utilities that enforce:
- Single-decode enforcement (decode at most twice)
- Prevention of auth page redirects
- Same-origin relative paths only
- Strip fragments (#...), collapse //, remove nested ?next=...
- Blocklist enforcement for auth paths
"""

import logging
import re
from urllib.parse import parse_qs, unquote, urlencode, urlparse

logger = logging.getLogger(__name__)

# Default fallback path
DEFAULT_REDIRECT = "/dashboard"

# Auth paths that should never be redirected to (exact match or prefix match)
BLOCKLIST_PATHS = {"/login", "/v1/auth", "/google", "/oauth"}


def safe_decode_url(url: str, max_decodes: int = 2) -> str:
    """
    Safely decode a URL-encoded string at most max_decodes times.

    Args:
        url: URL string to decode
        max_decodes: Maximum number of decode operations (default: 2)

    Returns:
        Decoded URL string
    """
    decoded = url
    previous = url

    for _ in range(max_decodes):
        try:
            previous = decoded
            decoded = unquote(decoded)

            # Stop if no change (no more encoding layers)
            if decoded == previous:
                break
        except Exception:
            # If decoding fails at any point, use the last successfully decoded version
            decoded = previous
            break

    return decoded


def is_blocklisted_path(path: str) -> bool:
    """
    Check if path is blocked for redirects.

    Uses exact match or prefix match for blocklisted paths.

    Args:
        path: Path to check

    Returns:
        True if path is blocklisted
    """
    if not path:
        return False

    # Check exact matches and prefix matches
    for blocked_path in BLOCKLIST_PATHS:
        if path == blocked_path or path.startswith(blocked_path + "/"):
            return True

    return False


def sanitize_next_path(raw: str | None) -> str:
    """
    Sanitize a redirect path to prevent open redirects and nesting loops.

    Rules enforced:
    - Relative-only targets; reject absolute/protocol-relative (http…, //…) → fallback
    - Double-decode max; stop after stability or 2 rounds
    - Remove next from any query string in the final path
    - Blocklist: exact match or prefix match for /login, /v1/auth, /google, /oauth
    - Normalize: drop fragments, collapse repeated slashes (preserve leading /)

    Args:
        raw: Raw path from user input (query param, form, etc.)

    Returns:
        Sanitized path that starts with / and is safe for redirects
    """
    if not raw or not isinstance(raw, str):
        return DEFAULT_REDIRECT

    path = raw.strip()
    if not path:
        return DEFAULT_REDIRECT

    try:
        # Step 1: Safe URL decoding (at most twice)
        path = safe_decode_url(path, max_decodes=2)

        # Step 2: Reject absolute URLs to prevent open redirects
        if path.startswith(("http://", "https://")):
            logger.warning("Rejected absolute URL redirect: %s", path)
            return DEFAULT_REDIRECT

        # Step 3: Reject protocol-relative URLs (starting with // but not ///)
        if path.startswith("//") and not path.startswith("///"):
            logger.warning("Rejected protocol-relative URL redirect: %s", path)
            return DEFAULT_REDIRECT

        # Step 4: Ensure path starts with /
        if not path.startswith("/"):
            logger.warning("Rejected non-relative path redirect: %s", path)
            return DEFAULT_REDIRECT

        # Step 5: Strip fragments (#...)
        if "#" in path:
            path = path.split("#")[0]

        # Step 6: Remove any nested ?next=... parameters
        if "?" in path:
            parsed = urlparse(path)
            query_params = parse_qs(parsed.query)

            # Remove any next parameters
            if "next" in query_params:
                del query_params["next"]

            # Reconstruct path without next params
            if query_params:
                new_query = urlencode(query_params, doseq=True)
                path = f"{parsed.path}?{new_query}"
            else:
                path = parsed.path

        # Step 7: Prevent redirect loops by blocking auth-related paths
        if is_blocklisted_path(path):
            logger.warning("Rejected blocklisted path redirect: %s", path)
            return DEFAULT_REDIRECT

        # Step 8: Normalize redundant slashes
        path = re.sub(r"/+", "/", path)

        # Step 9: Basic path validation (no .. traversal)
        if ".." in path:
            logger.warning("Rejected path traversal redirect: %s", path)
            return DEFAULT_REDIRECT

        return path

    except Exception as e:
        logger.error("Error sanitizing redirect path %s: %s", raw, e)
        return DEFAULT_REDIRECT
